### PR TITLE
[Refactor] automated dandi upload improvements

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
Previously I had mentioned it might be a tad messy to directly use the dandi library to perform these operations, but after messing around with it some more it's really just a lack of documentation that made it hard to figure out how to properly call the functions directly.

This fixes it and removes any need  for subprocesses, which provides enhanced stdout display as well as fewer pid issues.

Working like a charm so far.

Also a slight restructuring to where exactly the `wait` occurs during the process deploy/kill cycle. No change to behavior, just makes code a bit simpler to read.

Oh, also includes an important fix regarding forced inclusion of `session_id` in dandi file name; when using this function to upload only a single NWBFile, the default usage of `dandi organize` will only use the `subject_id` to name the file, since that is all that is necessary to make it unique. I've requested they add this as a general feature of `organize`, but until they add it I'm doing manual renaming to fix it for now.

### CODE REVIEW

File: src/nwb_conversion_tools/tools/data_transfers.py [hunk 1]
```diff
@@ -8,9 +8,14 @@
...
 from shutil import rmtree
 from time import sleep, time
+from tempfile import mkdtemp
 
 import psutil
 from tqdm import tqdm
+from pynwb import NWBHDF5IO
+from dandi.download import download
+from dandi.organize import organize
+from dandi.upload import upload
```
Comment (h-mayorquin) [inline src/nwb_conversion_tools/tools/data_transfers.py:19]: Dandi chose rather generic names for their functions. As this whole module is for data transfer maybe we can alias them (i.e `import as dandi_download`)?

File: src/nwb_conversion_tools/tools/data_transfers.py [hunk 2]
```diff
@@ -329,60 +326,52 @@ def dandi_upload(
...
-        Dictionary used to specify mazimum timeout of each individual process in the DANDI upload.
-        Keys must be from ['validate', 'download', 'organize', 'upload'].
-        The value of each key is number of seconds to wait before forcibly terminating.
-        Set any value to None to wait until the process completes, however long that may be.
-        The default is for all processes except organize to wait until full completion
-        (in testing, organize subprocess doesn't seem to want to wait successfully).
+    cleanup : bool, optional
+        Whether or not to remove the dandiset folder path and nwb_folder_path.
+        Defaults to False.
     """
```
Comment (h-mayorquin) [inline src/nwb_conversion_tools/tools/data_transfers.py:332]: Way simpler, cool!

File: src/nwb_conversion_tools/tools/data_transfers.py [hunk 3]
```diff
@@ -329,60 +326,52 @@ def dandi_upload(
...
-    try:
-        rmtree(path=nwb_folder_path.parent / dandiset_id)
-        rmtree(path=nwb_folder_path)
-    except PermissionError:
-        warn("Unable to clean up source files and dandiset! Please manually delete them.")
+    download(urls=dandiset_url, output_dir=str(dandiset_folder_path), get_metadata=True, get_assets=False)
+    assert dandiset_path.exists(), "DANDI download failed!"
+
+    organize(paths=str(nwb_folder_path), dandiset_path=str(dandiset_path))
+    organized_nwbfiles = dandiset_path.rglob("*.nwb")
```
Comment (h-mayorquin) [inline src/nwb_conversion_tools/tools/data_transfers.py:349]: Do you need `rglob` here, would not a simple `glob` suffice?

File: src/nwb_conversion_tools/tools/data_transfers.py [hunk 4]
```diff
@@ -329,60 +326,52 @@ def dandi_upload(
...
+            dandi_stem_split.insert(1, f"ses-{session_id}")
+            corrected_name = "_".join(dandi_stem_split) + ".nwb"
+            organized_nwbfile.rename(organized_nwbfile.parent / corrected_name)
+    organized_nwbfiles = dandiset_path.rglob("*.nwb")
+    # The above block can be removed once they add the feature
+
+    assert len(list(dandiset_path.iterdir())) > 1, "DANDI organize failed!"
+
+    dandi_instance = "dandi-staging" if staging else "dandi"
+    upload(paths=[str(x.absolute()) for x in organized_nwbfiles], dandi_instance=dandi_instance)
```
Comment (h-mayorquin) [inline src/nwb_conversion_tools/tools/data_transfers.py:369]:
```suggestion
    paths = [str(nwbfile_path.absolute()) for nwbfile_path in organized_nwbfiles]
    upload(paths=paths, dandi_instance=dandi_instance)
```

**Review by h-mayorquin — COMMENTED** [review]
Seems great, maybe it would be a good idea to add comments to separate the three main areas:
download, organize and upload to separate the sections logically in this function.

File: src/nwb_conversion_tools/tools/data_transfers.py (hunk 3 above)
Comment (CodyCBakerPhD) [reply src/nwb_conversion_tools/tools/data_transfers.py:349]: organization goes into subfolders, e.g.,

```
dandiset_folder / dandiset_id / sub-1 / sub-1_ses-2_ecephys.nwb
dandiset_folder / dandiset_id / sub-2 / sub-2_ses-3_ecephys.nwb
```

File: src/nwb_conversion_tools/tools/data_transfers.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply src/nwb_conversion_tools/tools/data_transfers.py:19]: Good idea, added

File: src/nwb_conversion_tools/tools/data_transfers.py (hunk 3 above)
Comment (h-mayorquin) [reply src/nwb_conversion_tools/tools/data_transfers.py:349]: Thanks, I was curios, what does each sub-folder corresponds to?

**Review by h-mayorquin — DISMISSED** [review]

File: src/nwb_conversion_tools/tools/data_transfers.py (hunk 3 above)
Comment (CodyCBakerPhD) [reply src/nwb_conversion_tools/tools/data_transfers.py:349]: 'sub' is dandi shorthand for 'subject'

**Review by h-mayorquin — APPROVED** [review]

**Pull request merged**